lib/Parse/ParseStmt.cpp

```diff
@@ -364,10 +364,29 @@ ParserStatus Parser::parseBraceItems(SmallVectorImpl<ASTNode> &Entries,
     const bool IsAtStartOfLineOrPreviousHadSemi =
         PreviousHadSemi || Tok.isAtStartOfLine();
     if (!IsAtStartOfLineOrPreviousHadSemi) {
-      SourceLoc EndOfPreviousLoc = getEndOfPreviousLoc();
-      diagnose(EndOfPreviousLoc, diag::statement_same_line_without_semi)
-        .fixItInsert(EndOfPreviousLoc, ";");
-      // FIXME: Add semicolon to the AST?
+      Token previousToken = Lexer::getTokenAtLocation(SourceMgr, PreviousLoc);
+      Optional<tok> stmtTokMatch = None;
+      if (previousToken.getKind() == tok::identifier) {
+        SourceLoc startOfLineLocation = L->getLocForStartOfLine(SourceMgr, previousToken.getLoc());
+        SourceLoc locationOfTokenAtStartOfLine = Lexer::getTokenAtLocation(SourceMgr, startOfLineLocation).getLoc();
+        if (previousToken.getLoc() == locationOfTokenAtStartOfLine) {
+          stmtTokMatch = closestMatchingStmtTokenForIdentifier(previousToken.getText());
+        }
+      }
+      if (stmtTokMatch) {
+        ParserPosition parserPosition = ParserPosition(L->getStateForBeginningOfToken(previousToken), PreviousLoc);
+        backtrackToPosition(parserPosition);
```

Comment:
I don’t think you have to backtrack.  You only need the location of the typo’d keyword.

Reply:
Currently I used backtrack to update previous token with our corrected token kind and continue parsing as there were no error. 
Maybe there are better way to recover there for example just skip tokens until we met end of statement.
Do we have any other options @rintaro ?

Reply:
Unfortunately,`setKind()` to a keyword kind is not not an option because we have an invariant in libSyntax where keyword tokens have always have the fixed text.

An option (which @CodaFi is suggesting)  is to make `parseStmt***` functions receive keyword location like
```cpp
ParserResult<Stmt> Parser::parseStmtReturn(SourceLoc tryLoc, SourceLoc ReturnLoc) {
  if (ReturnLoc.isInvalid())
    ReturnLoc = consumeToken(tok::kw_return);

  ...
}
```
Then directly call them from this recovery logic. But that's probably not simple changes considering `TopLevelCodeDecl` handling.

So, I recommend to skip to the next statement, decl, or '}' using `skipUntilDeclStmtRBrace()`. Downside of this is that we loose the diagnostics and semantic syntax highlighting until the next statement. But I think that's better than lousy diagnostics.

Reply:
Hi, totally agree. I also thought that approach with backtracking and setting kind is too unnatural. Related to skipUntilDeclStmtRBrace(). If we start using such diagnostics there are case like this.

Reply:
I think the implementation of `skipSingle()` is *not* wrong. This is for example:
```swift
func foo() {
  if /*HERE*/ foo {
    return 1
  }
  return 2
}
```
When skipping from `/*HERE*/` to the next statement, we don't care about the contents of the `if` statement. We just want to skip to `return 1` at line 5.

> Introduce function possibleStartOfStmt() and update skipUntilDeclStmtRBrace.

Could you elaborate on this? How will `possibleStartOfStmt()` be different from `isStartOfStmt()`?

Reply:
Hi, @rintaro, in case of
``` swift
func bar() {
    let i = 5
    /*HERE*/ gaurd i > 5 else { return }
    gaurd i > 5 else { return }
}
```
We are starting from /*HERE*/ we find out that there are typo error and then we try to skip till the next statement.
We are skipping brackets and then we meet second guard but it is not a start of statement (because it is wrong typed), so we are just skipping until we met correct start of statement (like guard, while without error in typing).
My question is should we consider such case, when there are more than 1 statement in a row typed wrong ? 
If yes, then possibleStartOfStmt() will also try to typo statement, if it is for example gaurd we are matching guard and not skipping it.

Reply:
I don't think it's worth it. For now, let's use `skipUntilDeclStmtRBrace()` as is. In your test cases, just guard them by some other decl or statement like:
```swift
    gaurd i > 5 else { return }
    func parserSync() {}
    i i > 5 else { return }
```
or put them in dedicated braces like:
```swift
    do {
      gaurd i > 5 else { return }
    }
    do {
      i i > 5 else { return }
    }
```

Reply:
Ok, agree. Lets keep it simple.

Reply:
Hi, @rintaro after investigation I find out there simple usage of skipUntilDeclStmtRBrace() is not sufficient for cases like
```swift
// SR-8510
func foo() -> Int? { return 0 }
func bar() {
  gaurd /*HERE*/ let x = foo() else { return }  // expected-error {{consecutive statements}} // expected-note {{did you misspell 'guard'}} {{3-8=guard}}
  print(x)
}
```
We will stop here and then emit a bunch of unrelated errors. Are there any workarounds around this, if we are trying to use skipUntilDeclStmtRBrace() ?
Should we consider introducing another skipping function ?
```swift
void Parser::skipUntilStmtRBrace() {
  while (Tok.isNot(tok::eof, tok::r_brace, tok::pound_endif,
                 tok::pound_else, tok::pound_elseif,
                 tok::code_complete) &&
       !isStartOfStmt())
  skipSingle();
}
```

Reply:
For now, how about just skipping the first `Tok.isAny(tok::kw_let, tok::kw_var)` in this recovery logic? like:
```
if (Tok.isAny(tok::kw_let, tok::kw_var))
  consumeToken();
skipUntilDeclStmtRBrace()
```
It's not great for something like:
```swift
gurad
  let a = someOpt1,
  let b = someOpt2
else {
}
```
But I think supporting single `let`/`var` is good enough because this is just a recovery.

Reply:
Hi, @rintaro updated pull request. Currently recovery logic will work only if next token is let/var and wrong typed statement is on the beginning of the line. Also updated tests.